### PR TITLE
perf: yolox acceleration

### PR DESCRIPTION
**Pull request by manato**
## Description

This PullReq integrates performance improvement on YOLOX into autoware.universe.
For finer grain control of TensorRT, including switching calibration algorithm used for int8 quantization, this PullReq also modifies `tensorrt_common`.

This PullReq includes the following improvements:
- The series of pre-processing functions that are performed before YOLOX inference have been accelerated using CUDA
- Additional calibration algorithms for int8 quantization have been supported
  - Especially in terms of YOLOX, we observed a significant accuracy drop if we used `EntropyV2` calibration algorithm for int8 quantization. Therefore, we've added support for a more suitable algorithm (i.e., `MinMax`). Because suitable calibration algorithms may differ among network models, we've also added `Legacy` to secure availability.
- We've also added some utility functions, such as dumping network structure into a file and profile per-layer information

## Tests performed

`ros2 launch tensorrt_yolox yolox.launch.xml precision:=int8 calibration_image_list_path:=<path_to_file_that_contains_image_list>`
(This PullReq is tested using a rosbag that contains `/sensing/camera/camera0/image_rect_color/compressed`. Also, engine file conversion from ONNX may take several minutes)

After executing the above command, the node named `/tensorrt_yolox` will run:
- Subscribe: `/sensing/camera/camera0/image_rect_color`
- Publish:
  - `/perception/object_recognition/detection/rois0`
  - `/tensorrt_yolox/out/image` (used for visual confirmation)

## Effects on system behavior

Not applicable.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp [hunk 1]
```diff
@@ -92,18 +123,56 @@ using TrtUniquePtr = std::unique_ptr<T, InferDeleter<T>>;
...
-    std::unique_ptr<nvinfer1::IInt8EntropyCalibrator2> calibrator = nullptr,
+    std::unique_ptr<nvinfer1::IInt8Calibrator> calibrator = nullptr,
     const BatchConfig & batch_config = {1, 1, 1}, const size_t max_workspace_size = (16 << 20),
+    const BuildConfig & buildConfig = BuildConfig(),
     const std::vector<std::string> & plugin_paths = {});
 
+  /**
+   * @brief Load TensorRT engine
+   * @param[in] engine_file_path path for a engine file
+   * @return flag for whether loading are succedded or failed
```
Comment (yukkysaito) [inline common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp:154]:
```suggestion
   * @return flag for whether loading are succeeded or failed
```

File: common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp [hunk 2]
```diff
@@ -28,50 +28,81 @@ namespace fs = ::std::filesystem;
...
+  // type for calibration
+  std::string calib_type_str;
+
+  // DLA core ID that the process uses
+  int dla_core_id;
+
+  // flag for partial quanitzation in first layer
+  bool quantize_first_layer;  // For partial quantization
+
+  // flag for partial quanitzation in last layer
```
Comment (yukkysaito) [inline common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp:56]:
```suggestion
  // flag for partial quantization in last layer
```

File: common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp [hunk 3]
```diff
@@ -28,50 +28,81 @@ namespace fs = ::std::filesystem;
...
 {
-public:
-  Logger() : Logger(Severity::kINFO) {}
+  // type for calibration
+  std::string calib_type_str;
+
+  // DLA core ID that the process uses
+  int dla_core_id;
+
+  // flag for partial quanitzation in first layer
```
Comment (yukkysaito) [inline common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp:53]:
```suggestion
  // flag for partial quantization in first layer
```

File: perception/tensorrt_yolox/src/tensorrt_yolox_node.cpp [hunk 4]
```diff
@@ -32,23 +32,72 @@ TrtYoloXNode::TrtYoloXNode(const rclcpp::NodeOptions & node_options)
...
-  float nms_threshold = declare_parameter("nms_threshold", 0.7);
+  auto declare_parameter_with_description =
+    [this](std::string name, auto default_val, std::string description = "") {
+      auto param_desc = rcl_interfaces::msg::ParameterDescriptor{};
+      param_desc.description = description;
+      return this->declare_parameter(name, default_val, param_desc);
+    };
+
+  std::string model_path = declare_parameter_with_description("model_path", "");
+  std::string label_path = declare_parameter_with_description("label_path", "");
```
Comment (wep21) [inline perception/tensorrt_yolox/src/tensorrt_yolox_node.cpp:43]: Is it necessary to use `declare_parameter_with_description` when the description is empty?

File: perception/tensorrt_yolox/include/tensorrt_yolox/tensorrt_yolox.hpp [hunk 5]
```diff
@@ -51,25 +51,74 @@ struct GridAndStride
...
+   * @param[in] height original image height
+   * @warning if we don't allocate buffers using it, "preprocess_gpu" allocates buffers at the
+   * beginning
+   */
+  void init_preproces_buffer(int width, int height);
+
+  /**
+   * @brief output TensorRT profiles for each layer
+   */
+  void print_profiling(void);
```
Comment (yukkysaito) [inline perception/tensorrt_yolox/include/tensorrt_yolox/tensorrt_yolox.hpp:109]: Although lower_camel is recommended by Autoware, the function naming conventions should be consistent within the file.

**Comment by wep21** [comment]
@manato Could you show the result of the performance improvement?

File: common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp (hunk 1 above)
Comment (manato) [reply common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp]: @yukkysaito 
Thank you for pointing this out. I correct the spelling at 3f059adf7a831b959038a6380497ffbb1feea5e9

File: common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp (hunk 3 above)
Comment (manato) [reply common/tensorrt_common/include/tensorrt_common/tensorrt_common.hpp]: same as https://github.com/autowarefoundation/autoware.universe/pull/3430#discussion_r1180202051

File: perception/tensorrt_yolox/src/tensorrt_yolox_node.cpp (hunk 4 above)
Comment (manato) [reply perception/tensorrt_yolox/src/tensorrt_yolox_node.cpp]: @wep21 
I appreciate your suggestion. I just used `declare_parameter_with_description` for consistency, but as you said, it looks somewhat weird to use `..._with_description` without a description. So I added simple descriptions for these arguments at 7f7a9f81c671a61d1635c28183bb4ebc25f3f726

File: perception/tensorrt_yolox/include/tensorrt_yolox/tensorrt_yolox.hpp (hunk 5 above)
Comment (manato) [reply perception/tensorrt_yolox/include/tensorrt_yolox/tensorrt_yolox.hpp]: @yukkysaito 
Thank you for your detailed review. To keep the naming rule consistent, I renamed some functions at d3601b9768b388c9baf8fbdf4d24539da476da8c

**Comment by manato** [comment]
@yukkysaito @wep21 
I really appreciate your detailed review. And sorry for my carelessness regarding misspellings and inconsistent namings. I fixed those problems at the succeeded commits. It should be noted that this PR restricts the TensorRT version. **Only TensorRT ver8.0 and higher will be supported** after merging this PR. 

Regarding performance improvement, I measured the execution time on the two environments:
- Desktop
  - CPU: Intel Core i9
  - GPU: RTX3070
  - CUDA: 11.6
  - TensorRT: 8.6.0
- Jetson (RQX-58G)
  - CPU: ARMv8
  - GPU: AGX Xavier
  - CUDA: 10.2
  - TensorRT: 8.0.1

The attachment depicts the results. We can see the modifications by this PR improve execution speed in both environments.  
elapsed_comparison


Let me know your thoughts.

**Review by yukkysaito — APPROVED** [review]
LGTM

**Comment by wep21** [comment]
@manato This is just a question. Why does this PR improve inference part of Desktop more than that of AGX Xavier?

**Comment by manato** [comment]
@wep21 
Before this PR's modification, input images are copied from CPU to GPU via pageable memory allocation and synchronous memory copy (i.e., `cudaMemcpy`). We modified this copy to use page-locked memory allocation and asynchronous memory copy ( `cudaMemcpyAsync`). I investigated the profile results roughly and found that this modification reduced the elapsed time by memory-copy-related operations. 

@dan-dnn 
If you have any additional insights about the question from @wep21, could you share them with us?

**Comment by manato** [comment]
@wep21 CC: @dan-dnn 
I re-checked the results and it revealed that the values for "Before, fp16, desktop" was measured using fp32.
My apologies for the mistake. I'm attaching the corrected result. The inference time was slightly improved because of the memory handling described above but basically almost similar.

elapsed_comparison

**Comment by manato** [comment]
@wep21 
Sorry to annoy you. This is a friendly reminder regarding the updates made after your last comment. If you have additional questions, please feel free to let me know. If your concerns are gone, I would appreciate it if you approve this PR. Thank you!

**Review by wep21 — APPROVED** [review]

**Comment by wep21** [comment]
@manato Thank you. The new graph you shared makes sense.
This is an additional question. I guess the original implementation of yolox doesn't use letter box resize, but this PR introduces letter box resize. Is there any reason to use letter box resize?

**Comment by manato** [comment]
@wep21 CC: @dan-dnn 
Thank you for the prompt response.
Regarding letter box resize, both [the original implementation](https://github.com/Megvii-BaseDetection/YOLOX/blob/main/yolox/data/data_augment.py#L154) and [autoware implementation](https://github.com/autowarefoundation/autoware.universe/blob/main/perception/tensorrt_yolox/src/tensorrt_yolox.cpp#L109) seem to use letter box resize. This PR just follows that fashion!

**Comment by wep21** [comment]
@manato Sorry, you're right. That was my misunderstanding.

**Comment by manato** [comment]
@yukkysaito @wep21 @miursh @dan-dnn 
Since all concerns seem to be cleared, I merged this PR. Thank you, guys, for all your cooperation.

**Pull request merged**